Crud/CrudServiceClient.cs

```diff
@@ -0,0 +1,270 @@
+        private const string ExportLiteral = "export";
+        private const string ApiSecretHeaderKey = "secret";
+
+
+        public CrudServiceClient(Dictionary<string, string> miaHeaders, string apiPath = default(string),
+            string apiSecret = default(string), int crudVersion = default(int))
+        {
+            ApiPath = apiPath;
+            CrudVersion = crudVersion;
+
+            AddSecretHeader(apiSecret);
+            AddMiaHeaders(miaHeaders);
+        }
+
+        private static void AddMiaHeaders(Dictionary<string, string> miaHeaders)
+        {
+            foreach (var (key, value) in miaHeaders)
+            {
+                Client.DefaultRequestHeaders.Remove(key);
+                Client.DefaultRequestHeaders.Add(key, value);
+            }
+        }
+
+        private static void AddSecretHeader(string apiSecret)
+        {
+            if (apiSecret != default(string) && !Client.DefaultRequestHeaders.Contains(ApiSecretHeaderKey))
+            {
+                Client.DefaultRequestHeaders.Add(ApiSecretHeaderKey, apiSecret);
+            }
+        }
+
+        static CrudServiceClient()
+        {
+            Client = new HttpClient();
+        }
+
+        private string BuildPath(string collectionName)
+        {
+            var basePath = ApiPath;
+            var suffix =
+                CrudVersion != default(int)
+                    ? $"/{VersionLiteral}{CrudVersion.ToString()}/{collectionName}"
+                    : $"/{collectionName}";
+            return basePath + suffix;
+        }
+
+        private Uri BuildUrl(string path, string query)
+        {
+            var uriBuilder = new UriBuilder(path) {Query = query};
+            return uriBuilder.Uri;
+        }
+
+        private string BuildQueryString(IReadOnlyCollection<KeyValuePair<string, string>> query)
+        {
+            return query == null ? new QueryBuilder(DefaultQuery).ToQueryString().ToString() : new QueryBuilder(DefaultQuery.Concat(query)).ToQueryString().ToString();
+        }
+
+        private static string GetCollectionName<T>()
+        {
+            return typeof(T).GetTypeInfo().GetCustomAttribute<CollectionName>()?.Value;
+        }
+
+        private async Task<HttpResponseMessage> SendAsyncRequest(HttpMethod method, string path,
+            string queryString = "",
+            string body = "")
+        {
+            try
+            {
+                var httpRequestMessage = new HttpRequestMessage
+                {
+                    Method = method,
+                    RequestUri = BuildUrl(path, queryString),
+                    Content = new StringContent(body, Encoding.UTF8, MediaTypeNames.Application.Json)
+                };
+
+                var response = await Client.SendAsync(httpRequestMessage);
+                return response;
+            }
+            catch (HttpRequestException e)
+            {
+                // TODO log error
```

Comment:
Logging should be delegated to who uses the library. We better throw a custom CRUD-related exception.

Reply:
Done

Reply:
I don't get it.. There's is still the TODO and you're actually returning an http response with status code 500 but if something went wrong the CRUD service provides a proper status code, we should pass over that status code (and body).

I guess this catch block is executed whenever the HTTP request fails and thus we've no status code at all, why faking a 500? Let's return a proper error

Reply:
Do you think it's ok to throw an exception this way?